context/src/junit/bindings.rs

```diff
@@ -31,6 +32,178 @@ pub struct BindingsReport {
+                    .iter()
+                    .filter(|tc| tc.status.status == BindingsTestCaseStatusStatus::Skipped)
+                    .count();
+                let failures = testcases
+                    .iter()
+                    .filter(|tc| tc.status.status == BindingsTestCaseStatusStatus::NonSuccess)
+                    .count();
+                let timestamp = testcases.iter().map(|tc| tc.timestamp.unwrap_or(0)).max();
+                let timestamp_micros = testcases
+                    .iter()
+                    .map(|tc| tc.timestamp_micros.unwrap_or(0))
+                    .max();
+                let time = testcases.iter().map(|tc| tc.time.unwrap_or(0.0)).sum();
+                report_time += time;
+                report_failures += failures;
+                report_tests += tests;
+                BindingsTestSuite {
+                    name,
+                    tests,
+                    disabled,
+                    errors: 0,
+                    failures,
+                    timestamp,
+                    timestamp_micros,
+                    time: Some(time),
+                    test_cases: testcases,
+                    properties: vec![],
+                    system_out: None,
+                    system_err: None,
+                    extra: HashMap::new(),
+                }
+            })
+            .collect();
+        let (name, timestamp, timestamp_micros) = match uploader_metadata {
+            Some(t) => (
+                t.origin,
+                Some(t.upload_time.unwrap_or_default().seconds),
+                Some(t.upload_time.unwrap_or_default().nanos as i64 * 1000),
+            ),
+            None => ("Unknown".to_string(), None, None),
+        };
+        BindingsReport {
+            name,
+            test_suites,
+            time: Some(report_time),
+            uuid: None,
+            timestamp,
+            timestamp_micros,
+            errors: 0,
+            failures: report_failures,
+            tests: report_tests,
+        }
+    }
+}
+
+impl From<TestCaseRun> for BindingsTestCase {
+    fn from(
+        TestCaseRun {
+            name,
+            parent_name,
+            classname,
+            started_at,
+            finished_at,
+            status,
+            status_output_message,
+            id,
+            file,
+            line,
+            attempt_number,
+        }: TestCaseRun,
+    ) -> Self {
+        let (timestamp, timestamp_micros) = match started_at {
+            Some(started_at) => (
+                Some(started_at.seconds),
+                Some(started_at.nanos as i64 * 1000),
+            ),
+            None => (None, None),
+        };
+        let time = match (started_at, finished_at) {
+            (Some(started_at), Some(finished_at)) => Some(
+                (finished_at.seconds - started_at.seconds) as f64
+                    + (finished_at.nanos - started_at.nanos) as f64 / 1_000_000_000.0,
```

Comment:
Is the intention here to just handle when one is set but the other is not? Wouldn't this double the actual time if both are set? Feels brittle imo

Reply:
No? This is to generate the duration.

Reply:
If
`finished_at.seconds = 2`
`started_at.seconds = 1`
`finished_at.nanos = 2_000_000_000`
`started_at.nanos = 1_000_000_000`

this would give you a duration of `2` seconds, would it not?

Reply:
`nanos` captures the nanos relative to the second. We can enforce that it's never a full second, but that's how it's supposed to work with google timestamp .

Reply:
Ah, didn't know the relative part